### ISSUE TITLE
Kilter Board Party mode

### ISSUE THREAD

**Issue by marcodejongh** [lemeryfertitta/Climbdex#80]
A big (first world problem) we have when climbing with a bunch of friends on our kilter, is that only 1 person knows what problem is currently on the board. It would be super sick if we could create a kilter "party" similar to how Spotify etc have such features.

Initially, that party mode could just show the current problem, and if people are logged in they could tick it.
But if #79 is possible, more options become available, like putting new problems in the queue etc.

**Comment by lemeryfertitta** [comment]
Worth reading through https://github.com/lemeryfertitta/Climbdex/issues/45 which already has some discussion about this feature - @gardaholm had some similar ideas.

Just a heads-up that I don't feel like I really finished a solid initial implementation of the Bluetooth feature. From my limited testing, I think it still needs some stability and UI work, especially around reconnection, indication of existing connection, etc. Might be worth playing around with it for a bit to see what you think, as improving the initial implementation to be at least as good as the official apps might be the first step on the road to more advanced functionality.

I'd also suggest we break this down into more specific issues as you mentioned. I agree that a nice first step would just be a sharable link that takes you to the current "session." Then the features from the "session" can be improved with things like ticking and adding climbs to the queue.

**Comment by marcodejongh** [comment]
Yeah I think the desire for more advanced features already pushed me in the direction of needing to overhaul stuff. It's currently all DOM state and HTML without a library, which becomes really cumbersome when pushing state changes from other places. 
So armed with ChatGPT, I spiked converting some of the UI to React this evening, and I got a proof of concept of editing a board and sharing between multiple browser sessions.
The cool bit about webrtc is that it doesn't involve any backend changes, and since nobody would argue their kilter board session is mission critical that should be plenty sufficient.

**Comment by lemeryfertitta** [comment]
> the desire for more advanced features already pushed me in the direction of needing to overhaul stuff. It's currently all DOM state and HTML without a library, which becomes really cumbersome when pushing state changes from other places.

Makes sense. The original intent of the app was just to provide search by hold functionality, so a heavy JS framework felt like overkill at the time. The scope has started to increase as it has become clear that Climbdex could fully replace and improve upon the functionality of the existing apps, and I can see an argument for the need of something more.

> I spiked converting some of the UI to React this evening, and I got a proof of concept of editing a board and sharing between multiple browser sessions.

Nice! If you do end up with something that you'd like to contribute, I'd be happy to review and get it deployed. My only request is that we break the changes down into smaller components to avoid a massive PR that changes every file 😅. If there is a conversion of existing UI, let's keep that separate from new features, and ideally also separated by the different pages.

**Comment by gardaholm** [comment]
> A big (first world problem) we have when climbing with a bunch of friends on our kilter, is that only 1 person knows what problem is currently on the board. It would be super sick if we could create a kilter "party" similar to how Spotify etc have such features.

this would be a real game-changer. we also have the same first world problem here, as we tend to use the tb2 in groups. a session mode would make things so much more comfortable – even if there is just one host (which can select climbs and light climbs) and all others in the session can only view (and log) the current climb. would be a great addition.

**Comment by marcodejongh** [comment]
@lemeryfertitta Yeah sounds good, I've put it behind a /react endpoint, it's relatively close to complete now, just haven't had a lot of time lately to finish it

**Comment by marcodejongh** [comment]
Uploading Screen Recording 2024-09-13 at 14.35.00-1.mov…

**Comment by marcodejongh** [comment]
@lemeryfertitta @gardaholm  I have a proof of concept with basic session sharing up: https://www.boardsesh.com/climb/kilter/8/25
When session sharing, currently it works like a remote control, but it should work more like spotify jams. The top kilter board will be for searching a climb, the bottom floating bar for playlist control of what is currently on the board. This way it also avoids jarring accidentally climb switches when someones currently climbing, and will make it easy to undo such a mistake as the forward and back buttons will go through history not the current search.

The Jams feature uses WebRTC so doesnt have to go through the backend.

I'll have to clean the repo up some more and then I'll open source it. I also converted the backend to nextjs, so it's pretty much a rewrite at this point, but it still uses Boardlib to get all the data.

**Comment by gardaholm** [comment]
nice! would love to test it in a group session here, but I've only got the tension board here.
do you think its possible to add tension data, I guess it's only missing the bin/sync_db.sh?

but will also do some dry runs with the kilter setup

**Comment by marcodejongh** [comment]
@gardaholm I'm on holiday atm, so havent implemented the bluetooth code yet since I have no way of testing it. But I should be able to get tension board in there.

I'll have a stab at the bluetooth code too, probably the climbdex module for it is completely portable.

**Comment by lemeryfertitta** [comment]
@marcodejongh cool! The sharing feature is neat. If you wanted to contribute that back to Climbdex I'd certainly be happy to accept that PR, but I understand if you prefer working on your rewrite. If you do continue with the rewrite and end up reaching feature parity with Climbdex, as well as making the work open source, I'd be happy to redirect to your project at that point and potentially contribute as well.

**Comment by marcodejongh** [comment]
I'll definitely make it open source, and I'm keen to get other people involved. The only reason why its closed right now is that it's using the pro tie of the Vercel deployment stuff and I cant figure out how that works for public repository.

this has been my fun holiday side project, but start of October I'm back to regular life so wont have as much time as I do now. I'll make sure to get alternative deployments sorted before then.

**Comment by marcodejongh** [comment]
@lemeryfertitta @gardaholm it's open source now: https://github.com/marcodejongh/boardsesh, and I've added instructions on how to start up a developer environment. I've also added Tension: https://www.boardsesh.com/tension/10/6/12,13/40/list.
Currently I don't have a board selector so here's kilter HW: https://www.boardsesh.com/kilter/8/17/26,27/40/list and regular OG Kilterboard: https://www.boardsesh.com/kilter/1/10/1,20/40/list

Contributions more than welcome. Next week I'll record some architecture run throughs.

**Comment by lemeryfertitta** [comment]
Nice! I'm going to continue accepting contributions on Climbdex for now, as it seems to still have functionality/usability that  boardsesh does not, but I'll keep an eye on the repo. If it gets to the point where all of the Climbdex functionality/usability is replicated and improved upon over there, I'll look into redirecting to that repo and website.

As an aside, I would advise you not to copy and re-host the board images. I'm not a lawyer, but I think that trends closer to copyright infringement, which is why I simply link to the images hosted by the board apps.
